src/cards/polymorph.ts

```diff
@@ -0,0 +1,165 @@
+                // Add new unit to targets
+                addTarget(newUnit, state, underworld);
+              }
+            } else {
+              console.error("Polymorph failed to choose a new unit type.");
+            }
+          } else if (isPickup(target)) {
+            const pickup = target as IPickup;
+            // Don't polymorph purple portals
+            if (pickup.name == Pickup.PORTAL_PURPLE_NAME) continue;
+
+            let possiblePickupTypes = pickups.filter(p => isModActive(p, underworld) && p.name != pickup.name
+              && p.name != Pickup.PORTAL_PURPLE_NAME && p.name != Pickup.RECALL_POINT);
+
+            // We have to seed this to prevent multiplayer desync
+            const seed = seedrandom(`${underworld.turn_number} -${pickup.id} `);
+            const chosenPickupType = chooseOneOfSeeded(possiblePickupTypes, seed);
+            if (chosenPickupType) {
+              // Replace current pickup with new pickup
+              const newPickup = polymorphPickup(pickup, chosenPickupType, underworld, prediction);
+              if (newPickup) {
+                // Visual aid only
+                if (prediction) {
+                  // TODO - Show new pickup?
+                }
+                // Cleanup old pickup and remove it from targets
+                Pickup.removePickup(pickup, underworld, prediction);
+                state.targetedPickups = state.targetedPickups.filter(p => p != pickup);
+                // Add new pickup to targets
+                addTarget(newPickup, state, underworld);
+              }
+            } else {
+              console.error("Polymorph failed to choose a new unit type.");
+            }
+          }
+        }
+      }
+
+      return state;
+    },
+  },
+};
+
+function polymorphUnit(fromUnit: Unit.IUnit, toUnitId: string, underworld: Underworld, prediction: boolean): Unit.IUnit | undefined {
+  const sourceUnit = allUnits[toUnitId];
+  if (!sourceUnit) {
+    console.error('Unit with id', toUnitId, 'does not exist. Have you registered it in src/units/index.ts?');
+    return undefined;
+  }
+
+  let unit: Unit.IUnit = Unit.create(
+    sourceUnit.id,
+    fromUnit.x,
+    fromUnit.y,
+    fromUnit.faction,
+    sourceUnit.info.image,
+    UnitType.AI,
+    sourceUnit.info.subtype,
+    { ...sourceUnit.unitProps, isMiniboss: fromUnit.isMiniboss, originalLife: fromUnit.originalLife },
+    underworld,
+    prediction
+  );
+  return unit;
+}
+
+function polymorphPickup(fromPickup: IPickup, toPickupSource: Pickup.IPickupSource, underworld: Underworld, prediction: boolean): IPickup | undefined {
+  const pickup = Pickup.create({ pos: fromPickup, pickupSource: toPickupSource, logSource: 'spawnPickup' }, underworld, prediction);
+  if (!prediction) {
+    setTimeout(() => {
+      playSFXKey('spawnPotion');
+      floatingText({ coords: fromPickup, text: toPickupSource.name });
+    }, 500);
+  }
+  return pickup;
+}
```

Comment:
Seems odd that the cleanup of the original pickup doesn't happen in this function.  Though I get it because you need to remove it from targets, but those two behaviors definitely belong together (imagine if this function were called in a different context).  I'd say inline the contents of this function for that reason.  Same for the above polymorphUnit.  I don't see any reason that they should be extracted as individual functions, particularly because they don't handle all the logic inherent in polymorphing.